### DISCUSSION TITLE
Why there is no deadlock?

### DISCUSSION THREAD

**Discussion by wddpfx** [eProsima/Fast-DDS#4455] [General]
In the file Fast-DDS\examples\cpp\dds\TypeLookupService\TypeLookupSubscriber.cpp, we create DataReader in a lambda, Which is registered in the function:register_remote_type. The picture shows detail:
![image_20](https://github.com/eProsima/Fast-DDS/assets/19266720/bfb14f27-56fc-4984-8b88-625b1ecc1d5e)

But the listener will be called when the subscriber receiver a message,  and when the message moudle process the data, the shared_lock is created:
![image_21](https://github.com/eProsima/Fast-DDS/assets/19266720/8649c2e4-f3b8-46f8-9e6b-d6abe19c3798)
This is reader lock,right?

And when we call the function "on_type_information_received" in listener, we will create a DataReader. And in the process,we will associate the reader endpoint with a MessageReceiver, But we will create a writer lock too, they are the same shared_mutex. why there is no deadlock.

![image_22](https://github.com/eProsima/Fast-DDS/assets/19266720/4e736f22-1781-4f4b-8938-fa5c6312bbdd)

May be they are not the same MessageReceiver? I think it should not be, both of them are udp transport.
Sorry my English is poor, can you understand me?

**Comment by MiguelCompany** [comment]
> May be they are not the same MessageReceiver?

Correct, they are not the same MessageReceiver. The first one is the MessageReceiver for the meta-traffic (i.e. discovery) listening port. The second one is for the user traffic listening port.

That said, we don't recommend creating endpoints inside callbacks, since the internal behavior could change in the future.

**Comment by wddpfx** [comment]
好的，收到了你的邮件，我会尽快查看。
